quantize.py

```diff
@@ -0,0 +1,242 @@
+
+    model.transformer.wte.to(working_device)
+    inps = []
+    for batch in sample_inputs:
+        inps.append(model.transformer.wte(batch[None].to(working_device)))
+    inps = torch.cat(inps, dim=0)
+    model.transformer.wte.to("cpu")
+    torch.cuda.empty_cache()
+
+    print("Starting to quantize blocks")
+    outs = torch.zeros_like(inps)
+
+    # better than relying on enumeration? originally the code bundled
+    # the two mlp fc layers
+    # we could automate this with a lot of hooks and another iteration
+    submodules_to_process = [
+        "attn.c_attn",
+        "attn.c_proj",
+        "mlp.c_fc1",
+        "mlp.c_fc2",
+        "mlp.c_proj",
+    ]
+
+    for i, block in enumerate(model.transformer.h):
+        block.to(working_device)
+
+        for name in submodules_to_process:
+            print(i, name, end=" ")
+            t0 = time.perf_counter()
+            print("collecting stats", end=" ")
+            sys.stdout.flush()
+            module = block.get_submodule(name)
+
+            gptq = GPTQQuantizer(
+                module,
+                bits=bits,
+                groupsize=groupsize,
+                actorder=(groupsize == -1),
+            )
+            handle = module.register_forward_hook(gptq.collect_input_stats)
+            for j in range(inps.size(0)):
+                outs[j : j + 1] = block(inps[j : j + 1])
+
+            handle.remove()
+
+            print("quantizing", end=" ")
+            sys.stdout.flush()
+            q_module, error = gptq.quantize()
+
+            # replace the linear module with the quantized module
+            pname, dname = name.rsplit(".", 1)
+            setattr(block.get_submodule(pname), dname, q_module)
+
+            # cleanup in an attempt to not run out of memory
+            del gptq
+            gc.collect()
+            torch.cuda.empty_cache()
+            t1 = time.perf_counter()
+            print(f"time {int(t1 - t0 + 0.5)}s quantization error {error:.1f}")
+
+        for j in range(inps.size(0)):
+            outs[j : j + 1] = block(inps[j : j + 1])
+
+        block.cpu()
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        # the outputs are the next block's inputs and we'll reuse the old inputs
+        inps, outs = outs, inps
+
+    model.transformer.ln_f.to(working_device)
+    for j in range(inps.size(0)):
+        outs[j : j + 1] = model.transformer.ln_f(inps[j : j + 1])
+    model.transformer.ln_f.to("cpu")
+    inps, outs = outs, inps
+
+    model.lm_head.to(working_device)
+    gptq = GPTQQuantizer(
+        model.lm_head,
+        bits=bits,
+        groupsize=groupsize,
+        actorder=(groupsize == -1),
+    )
+    handle = model.lm_head.register_forward_hook(gptq.collect_input_stats)
+    for j in range(inps.size(0)):
+        model.lm_head(inps[j : j + 1])
+    handle.remove()
+    q_module, error = gptq.quantize()
+    model.lm_head = q_module
+    model._original_module.lm_head = q_module  # bleh.
```

Comment:
@awaelchli Ideas for this? If I don't I get the "original" linear module.

Reply:
I think you may want to move

`model = fabric.setup_module(model)`
after
`llama_blockwise_quantization`

Otherwise, the model is wrapped already and that looks undesired here.

Reply:
I just opened https://github.com/Lightning-AI/lightning/issues/17285 with an idea to solve this so the user doesn't have to think about it.

Reply:
@t-vi good to merge, but do you want to try moving `model = fabric.setup_module(model)` after `llama_blockwise_quantization` as suggested before merging?

Reply:
What would be the difference to commenting out? The only step left after the quantization is to save the state dict.

Reply:
Probably no difference at this point. We are passing `fabric` to save_model_checkpoint` but I don't think we are leveraging any fabric-specific functionality here.

At this point I would just remove the use of fabric from `quantize.py` altogether, as it's not needed.

Reply:
OK